contracts/multisig-prover/src/execute.rs

```diff
@@ -121,29 +123,108 @@ pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractEr
         pub_keys.push(pub_key);
     }
 
-    let new_worker_set = WorkerSet::new(snapshot, pub_keys, env)?;
+    Ok((snapshot, participants.into_iter().zip(pub_keys).collect()))
+}
+
+pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    let (snapshot, participants) = get_workers_info(&deps, &env)?;
 
-    let cur_worker_set = CURRENT_WORKER_SET.load(deps.storage)?;
+    let new_worker_set = WorkerSet::new(participants.clone(), snapshot.quorum.into(), env)?;
+
+    let cur_worker_set = CURRENT_WORKER_SET.may_load(deps.storage)?;
+
+    // if no worker set, just store it and return
+    if cur_worker_set.is_none() {
+        CURRENT_WORKER_SET.save(deps.storage, &new_worker_set)?;
+
+        KEY_ID.save(deps.storage, &new_worker_set.hash().to_hex())?;
+
+        let key_gen_msg = multisig::msg::ExecuteMsg::KeyGen {
+            key_id: new_worker_set.id(),
+            snapshot,
+            pub_keys: participants
+                .into_iter()
+                .map(|(participant, pub_key)| {
+                    (
+                        participant.address.to_string(),
+                        (KeyType::ECDSA, <&[u8]>::from(&pub_key).into()),
+                    )
+                })
+                .collect(),
+        };
+
+        return Ok(Response::new().add_message(wasm_execute(
+            config.multisig,
+            &key_gen_msg,
+            vec![],
+        )?));
+    }
```

Comment:
We don't really even need the keygen execute msg at all anymore. All keygen is actually doing is storing the workerset in the multisig contract, so later on when we trigger signing sessions, we can just pass some id for the workerset, instead of the whole workerset. I would like to just remove keygen and store the workerset in the prover only, but I think that should be left to a later PR.